README.md
# XFI

XFI takes advantage of the features available in HTML5 to provide a better user experience for file inputs using the multiple attribute.

If you're looking for features like drag and drop, asynchronous uploads and progress bars then you might want to try something else.

XFI is a simple a jQuery plugin that can be easily integrated into your existing code.

Project status: **Active**

Current version: **1.1**

- [View a demo](http://alistairholt.co.uk/projects/xfi/demo)
- [Read the documentation](http://alistairholt.co.uk/projects/xfi#documentation)

## What XFI provides

- A consistent interface for file inputs using the multiple attribute across different browsers and platforms that is easily styleable with CSS
- Feedback to the user about selected file(s)
- Feedback to the user about file uploads taking place on form submission (Optional)

## What XFI doesn't provide

- Drag and drop file uploads
- Asynchronous file uploads
- Progress bars
- File type/size validation

## Documentation

Setting up XFI is really simple:

- Make sure jQuery is available in your page
- Add the XFI script to your page
- Set the multiple attribute and add a CSS class of '`xfi`' to any file inputs on which you want to enable XFI
- Call `$.XFI()` after the DOM has loaded

XFI only has two configuration options, both of which are optional:

**submitButtonFeeback**

(Boolean) Whether or not to use the form's submit button to provide feedback to the user about file uploads that are taking place on form submission. Defaults to true.

**debug**

(Boolean) Whether or not to put XFI into debug mode. In debug mode XFI will log to the JavaScript console if it is available. Defaults to false.

For example:

    $.XFI({
      submitButtonFeedback: false,
      debug: true
    });

The HTML generated by XFI is easily styleable via CSS.

XFI proxy input markup:

    <button class="xfi-input">Choose file(s)...</button>

File information list markup:

    <ul class="xfi-file-information-container">
      <li class="xfi-file-information">IMG_0001.jpg <i>(560 Kb)</i></li>
    </ul>

And finally, a CSS class of '`busy`' will be added to the form's submit button on submission if the submitButtonFeedback option is enabled.

## Browser Support

- Chrome 2+
- Safari 4+
- Firefox 3.6+
- Opera 11+
- Internet Explorer 10 (Developer Preview)

Internet Explorer 6-9 do not currently support the multiple attribute on file inputs, therefore XFI is not supported for Internet Explorer. Any XFI file inputs will fallback to being normal file input fields.

## Downloads

Current version - 1.1

- [Production](https://github.com/downloads/alistairholt/XFI/jquery.xfi-1.1.min.js.gz) (Minified and Gzipped - 4 Kb)
- [Development](https://github.com/downloads/alistairholt/XFI/jquery.xfi-1.1.js) (8 Kb)


## License

XFI is licensed under a MIT License

## Feedback

Feel free to tweet me any feedback on [@alistairholt](http://twitter.com/alistairholt).

## Contributions

Got a bug fix or something to add? Fork and send me a pull request on [GitHub](http://github.com/alistairholt/XFI).

If you just want to report an issue you've experience, [create one on GitHub](https://github.com/alistairholt/XFI/issues).

CHANGELOG.md
# 1.1

- Add CSS classes to generated markup to enable styling via CSS

# 1.0

- Initial release

jquery.xfi-1.0.js
(function ($) {
  $.XFI = function (options) {
    var self;

    self = this;

    self.settings = {
      submitButtonFeeback: true,
      debug: false
    };

    $.extend(self.settings, options);

    self.buttonText = "Choose File(s)...";

    self.log = function (message, object) {
      if (typeof console !== "undefined" && console !== null) {
        if (typeof object !== "undefined") {
          console.log("[XFI] " + message + ":");
          console.log(object);
        } else {
          console.log("[XFI] " + message);
        }
      }
    };

    self.generateUIDSegment = function () {
      return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    };

    self.generateUID = function () {
      return (self.generateUIDSegment() + "-" + self.generateUIDSegment());
    };

    self.humanizeFileSize = function (size) {
      var units, i;
      units = ['b', 'Kb', 'Mb', 'Gb', 'Tb', 'Pb', 'Eb', 'Zb', 'YB'];
      i = 0;
      while (size >= 1024) {
        size /= 1024;
        ++i;
      }
      return size.toFixed(1) + ' ' + units[i];
    };

    self.init = function () {
      if (self.settings.debug === true) {
        self.log("Initialized");
      }

      $("form:has(input[type='file'].xfi)").each(function () {
        var $form, $inputs;
        $form = $(this);
        $inputs = $form.find("input[type='file'].xfi");
        self.log("Found " + $inputs.length + " XFI file input(s)", $inputs);
        $inputs.each(function () {
          var $input, inputUID, $proxyInput, $cancel, $fileInfo;

          $input = $(this);

          // Generate an UID for this file input
          inputUID = self.generateUID();

          // Hide the existing file input
          // NOTE: We can't just set display to none
          // because Webkit stops us from proxying
          // events back to the input
          $input.css({
            position: "absolute",
            top: 0,
            left: "-1000em",
            width: 0,
            height: 0,
            visibility: "hidden"
          }).attr("data-xfi", inputUID);

          // Build a proxy file input
          $proxyInput = $("<button>", {
            type: "button",
            "class": "xfi-input",
            text: self.buttonText
          }).attr("data-xfi", inputUID);

          // Insert the proxy file input into the DOM
          $proxyInput.insertAfter($input);

          $cancel = $("<a>", {
            href: "",
            "class": "xfi-cancel",
            "data-xfi": inputUID,
            text: "Cancel",
            click: function (e) {
              var xfiUID, $input, $proxyInput, $fileInfo;
              xfiUID = $(this).data("xfi");
              // Reset the value for the original input
              $input = $("input[data-xfi='" + xfiUID + "'].xfi");
              $input.val("");
              // Reset the text for the proxy input
              $proxyInput = $("button[data-xfi='" + xfiUID + "']");
              $proxyInput.text(self.buttonText);
              // Clear the file information
              $fileInfo = $("ul[data-xfi='" + xfiUID + "']");
              $fileInfo.empty().hide();
              $(this).hide();
              if (self.settings.submitButtonFeeback) {
                $input.parents("form").unbind(".xfi");
              }
              e.preventDefault();
            }
          });

          $cancel.insertAfter($proxyInput).hide();
          $cancel.before(" ");

          // Build a UL to hold file information
          $fileInfo = $("<ul>").attr("data-xfi", inputUID);

          $fileInfo.insertBefore($proxyInput).hide();

          // Pass click events on the proxy file input
          // back to the orginal file input
          $proxyInput.bind("click", function (e) {
            var $proxyInput, xfiUID, $input;
            $proxyInput = $(this);
            xfiUID = $proxyInput.data("xfi");
            $input = $("input[data-xfi='" + xfiUID + "'].xfi");
            $input.click();

            e.preventDefault();
          });

          // Watch the original file input for changes
          $input.bind("change", function () {
            var $input, $proxyInput, proxyInputText, xfiUID, $fileInfo, files, $cancel;
            $input = $(this);
            xfiUID = $input.data("xfi");
            $proxyInput = $("button[data-xfi='" + xfiUID + "']");
            $fileInfo = $("ul[data-xfi='" + xfiUID + "']");
            $fileInfo.empty().hide();
            $cancel = $("a[data-xfi='" + xfiUID + "'].xfi-cancel");
            $cancel.show();
            files = $input[0].files;
            if (files.length > 0) {
              $(files).each(function () {
                var fileName, fileSize;
                // Mozilla File attributes
                //   name, size
                // WebKit File attributes
                //   name/fileName, size/fileSize
                // Opera File attributes
                //   name, size
                // Internet Explorer File attributes
                //   ?
                self.log("File chosen", this);
                fileName = this.name;
                fileSize = this.size;
                $("<li>", {
                  html: fileName + " <i>(" + self.humanizeFileSize(fileSize) + ")</i>"
                }).appendTo($fileInfo);
              });

              // Show file information
              $fileInfo.show();
              // Update proxy input
              proxyInputText = (files.length > 1) ? " files" : " file";

              $proxyInput.text(files.length + proxyInputText + " chosen. Change?");

              if (self.settings.submitButtonFeeback) {
                $input.parents("form").bind("submit.xfi", function () {
                  var $form, $xfiInputs, totalFileCount, fileOrFiles;
                  $form = $(this);
                  totalFileCount = 0;
                  $xfiInputs = $form.find("input.xfi");
                  $xfiInputs.each(function () {
                    totalFileCount += this.files.length;
                  });
                  if (totalFileCount > 0) {
                    fileOrFiles = (totalFileCount > 1) ? "files" : "file";

                    $form.find("[type='submit']:last")
                      .val("Uploading " + totalFileCount + " " + fileOrFiles + "... Please wait...")
                      .attr("disabled", "disabled");
                  }
                });
              }
            } else {
              if (self.settings.submitButtonFeeback) {
                $input.parents("form").unbind(".xfi");
              }
              $proxyInput.text(self.buttonText);
            }
          });
        });
      });
    };

    self.init();
  };
})(jQuery);


jquery.xfi-1.1.js
/*!
 * XFI 1.1
 * Release Date: 2011/12/14
 * http://alistairholt.co.uk/projects/xfi/
 *
 * Copyright (c) 2011 Alistair Holt, http://alistairholt.co.uk/
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

(function ($) {
  $.XFI = function (options) {
    var self;

    self = this;

    self.settings = {
      submitButtonFeeback: true,
      debug: false
    };

    $.extend(self.settings, options);

    self.buttonText = "Choose File(s)...";

    self.log = function (message, object) {
      if (typeof console !== "undefined" && console !== null) {
        if (typeof object !== "undefined") {
          console.log("[XFI] " + message + ":");
          console.log(object);
        } else {
          console.log("[XFI] " + message);
        }
      }
    };

    self.generateUIDSegment = function () {
      return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    };

    self.generateUID = function () {
      return (self.generateUIDSegment() + "-" + self.generateUIDSegment());
    };

    self.humanizeFileSize = function (size) {
      var units, i;
      units = ['b', 'Kb', 'Mb', 'Gb', 'Tb', 'Pb', 'Eb', 'Zb', 'YB'];
      i = 0;
      while (size >= 1024) {
        size /= 1024;
        ++i;
      }
      return size.toFixed(1) + ' ' + units[i];
    };

    self.init = function () {
      if (self.settings.debug === true) {
        self.log("Initialized");
      }

      $("form:has(input[type='file'].xfi)").each(function () {
        var $form, $inputs;
        $form = $(this);
        $inputs = $form.find("input[type='file'].xfi");
        self.log("Found " + $inputs.length + " XFI file input(s)", $inputs);
        $inputs.each(function () {
          var $input, inputUID, $proxyInput, $cancel, $fileInfo;

          $input = $(this);

          // Generate an UID for this file input
          inputUID = self.generateUID();

          // Hide the existing file input
          // NOTE: We can't just set display to none
          // because Webkit stops us from proxying
          // events back to the input
          $input.css({
            position: "absolute",
            top: 0,
            left: "-1000em",
            width: 0,
            height: 0,
            visibility: "hidden"
          }).attr("data-xfi", inputUID);

          // Build a proxy file input
          $proxyInput = $("<button>", {
            type: "button",
            "class": "xfi-proxy-input",
            text: self.buttonText
          }).attr("data-xfi", inputUID);

          // Insert the proxy file input into the DOM
          $proxyInput.insertAfter($input);

          $cancel = $("<a>", {
            href: "",
            "class": "xfi-cancel",
            "data-xfi": inputUID,
            text: "Cancel",
            click: function (e) {
              var xfiUID, $input, $proxyInput, $fileInfo;
              xfiUID = $(this).data("xfi");
              // Reset the value for the original input
              $input = $("input[data-xfi='" + xfiUID + "'].xfi");
              $input.val("");
              // Reset the text for the proxy input
              $proxyInput = $("button[data-xfi='" + xfiUID + "']");
              $proxyInput.text(self.buttonText);
              // Clear the file information
              $fileInfo = $("ul[data-xfi='" + xfiUID + "']");
              $fileInfo.empty().hide();
              $(this).hide();
              if (self.settings.submitButtonFeeback) {
                $input.parents("form").unbind(".xfi");
              }
              e.preventDefault();
            }
          });

          $cancel.insertAfter($proxyInput).hide();
          $cancel.before(" ");

          // Build a UL to hold file information
          $fileInfo = $("<ul>", {
            "class": "xfi-file-information-container"
          }).attr("data-xfi", inputUID);

          $fileInfo.insertBefore($proxyInput).hide();

          // Pass click events on the proxy file input
          // back to the orginal file input
          $proxyInput.bind("click", function (e) {
            var $proxyInput, xfiUID, $input;
            $proxyInput = $(this);
            xfiUID = $proxyInput.data("xfi");
            $input = $("input[data-xfi='" + xfiUID + "'].xfi");
            $input.click();

            e.preventDefault();
          });

          // Watch the original file input for changes
          $input.bind("change", function () {
            var $input, $proxyInput, proxyInputText, xfiUID, $fileInfo, files, $cancel;
            $input = $(this);
            xfiUID = $input.data("xfi");
            $proxyInput = $("button[data-xfi='" + xfiUID + "']");
            $fileInfo = $("ul[data-xfi='" + xfiUID + "']");
            $fileInfo.empty().hide();
            $cancel = $("a[data-xfi='" + xfiUID + "'].xfi-cancel");
            $cancel.show();
            files = $input[0].files;
            if (files.length > 0) {
              $(files).each(function () {
                var fileName, fileSize;
                self.log("File chosen", this);
                fileName = this.name;
                fileSize = this.size;
                $("<li>", {
                  "class": "xfi-file-information",
                  html: fileName + " <i>(" + self.humanizeFileSize(fileSize) + ")</i>"
                }).appendTo($fileInfo);
              });

              // Show file information
              $fileInfo.show();
              // Update proxy input
              proxyInputText = (files.length > 1) ? " files" : " file";

              $proxyInput.text(files.length + proxyInputText + " chosen. Change?");

              if (self.settings.submitButtonFeeback) {
                $input.parents("form").bind("submit.xfi", function () {
                  var $form, $xfiInputs, totalFileCount, fileOrFiles;
                  $form = $(this);
                  totalFileCount = 0;
                  $xfiInputs = $form.find("input.xfi");
                  $xfiInputs.each(function () {
                    totalFileCount += this.files.length;
                  });
                  if (totalFileCount > 0) {
                    fileOrFiles = (totalFileCount > 1) ? "files" : "file";

                    $form.find("[type='submit']:last")
                      .val("Uploading " + totalFileCount + " " + fileOrFiles + "... Please wait...")
                      .attr("disabled", "disabled")
                      .addClass("busy");
                  }
                });
              }
            } else {
              if (self.settings.submitButtonFeeback) {
                $input.parents("form").unbind(".xfi");
              }
              $proxyInput.text(self.buttonText);
            }
          });
        });
      });
    };

    self.init();
  };
})(jQuery);
